dataactbroker/handlers/accountHandler.py

```diff
@@ -867,28 +864,68 @@ def email_users(self, system_email, session):
+def perms_to_affiliations(perms):
+    """Convert a list of perms from MAX to a list of UserAffiliations. Filter
+    out and log any malformed perms"""
+    available_codes = {
+        cgac.cgac_code:cgac
+        for cgac in GlobalDB.db().session.query(CGAC)
+    }
+    for perm in perms:
+        components = perm.split('-PERM_')
+        if len(components) != 2:
+            logger.warning('Malformed permission: %s', perm)
+            continue
+
+        cgac_code, perm_level = components
+        perm_level = perm_level.lower()
+        if cgac_code not in available_codes or perm_level not in 'rws':
+            logger.warning('Malformed permission: %s', perm)
+            continue
+
+        yield UserAffiliation(
+            cgac=available_codes[cgac_code],
+            permission_type_id=PERMISSION_SHORT_DICT[perm_level]
+        )
 
 
-def grant_highest_permission(user, group_list, cgac_group):
-    """Find the highest permission within the provided cgac_group; set that as
-    the user's permission_type_id"""
-    user.cgac_code = cgac_group[-3:]
-    user.website_admin = False
-    permission_group = [g for g in group_list
-                        if g.startswith(cgac_group + "-PERM_")]
-    # Check if a user has been placed in a specific group. If not, deny access
-    if not permission_group:
-        user.permission_type_id = None
+def best_affiliation(affiliations):
+    """If a user has multiple permissions for a single agency, select the
+    best"""
+    by_agency = {}
+    affiliations = sorted(affiliations, key=attrgetter('permission_type_id'))
+    for affiliation in affiliations:
+        by_agency[affiliation.cgac] = affiliation
+    return by_agency.values()
+
+
+def set_max_perms(user, max_group_list):
+    """Convert the user group lists present on MAX into a list of
+    UserAffiliations and/or website_admin status.
+
+    Permissions are encoded as a comma-separated list of
+    {parent-group}-CGAC_{cgac-code}-PERM_{one-of-R-W-S}
+    or
+    {parent-group}-CGAC_SYS to indicate website_admin"""
+    prefix = CONFIG_BROKER['parent_group'] + '-CGAC_'
+    perms = [group_name[len(prefix):]
+             for group_name in max_group_list.split(',')
+             if group_name.startswith(prefix)]
+    if 'SYS' in perms:
+        user.affiliations = []
+        user.cgac_code = 'SYS'
+        user.website_admin = True
+        user.permission_type_id = PERMISSION_TYPE_DICT['writer']
     else:
-        perms = [perm[-1].lower() for perm in permission_group]
-        ordered_perms = sorted(
-            PERMISSION_MAP.items(), key=lambda pair: pair[1]['order'])
-        for key, permission in ordered_perms:
-            name = permission['name']
-            if key in perms:
-                user.permission_type_id = PERMISSION_TYPE_DICT[name]
-                break
+        affiliations = list(best_affiliation(perms_to_affiliations(perms)))
+
+        if not affiliations:
```

Comment:
PR https://github.com/fedspendingtransparency/data-act-broker-backend/pull/454 made some changes to allow folks that aren't in any groups whatsoever (so they don't need to be affiliated with an Agency) to access the broker with a null value for `permission_type_id` and `cgac_code`. This wouldn't be supported with this check for `if not affiliations` correct?